File: data-pack-app/engine/top_customers.py
"""
Top Customer Analysis tab generator.

Shows the top N customers ranked by latest-period ARR.

Three sections:
  1. ARR over time
  2. % YoY Growth
  3. % of Total

Rows:
  - Top N individual customers (looked up by rank)
  - Top N Total
  - Other Customers
  - Total ARR
  - Concentration memo (Top 3, 5, 10, 25)
"""
from openpyxl.utils import get_column_letter
from .utils import col_letter


TOP_N = 25
CONCENTRATION_TIERS = [3, 5, 10, 25]


def generate_top_customers_tab(wb, config, clean_sheet_name, clean_layout,
                               first_data_row, last_data_row):
    """Generate the Annual Top Customer Analysis tab."""
    sheet_name = "Annual Top Customer Analysis"
    ws = wb.create_sheet(sheet_name)

    num_dates = clean_layout['num_dates']
    num_attrs = clean_layout['num_attrs']

    # Column layout
    # B: rank number, C: Customer ID, D-F: attributes, G: cohort
    rank_num_col = 2  # B
    cust_id_col = 3   # C
    attr_start = 4    # D
    attr_end = attr_start + num_attrs - 1
    cohort_col = attr_end + 1

    # Section 1: ARR
    s1_start = cohort_col + 1
    s1_end = s1_start + num_dates - 1

    # Section 2: % YoY Growth (one fewer column than ARR)
    s2_start = s1_end + 2  # gap
    s2_end = s2_start + num_dates - 2  # n-1 growth periods

    # Section 3: % of Total
    s3_start = s2_end + 2  # gap
    s3_end = s3_start + num_dates - 1

    # Clean data references
    cdr_first = first_data_row
    cdr_last = last_data_row
    clean_rank_col = col_letter(clean_layout['rank'])
    clean_cust_id_col = col_letter(clean_layout['cust_id'])

    # --- Row 3: Units (A3=label, B3=formula) ---
    ws.cell(row=3, column=1, value="Units")
    ws.cell(row=3, column=rank_num_col, value="=Control!$C$4")
    units_cell = f"${col_letter(rank_num_col)}$3"

    # --- Row 5: Section headers ---
    metric_label = "ARR" if config.get("data_type", "arr") == "arr" else "Revenue"
    ws.cell(row=5, column=s1_start, value=metric_label)
    ws.cell(row=5, column=s2_start, value="% YoY Growth")
    ws.cell(row=5, column=s3_start, value="% of Total")

    # --- Row 6: Column headers ---
    ws.cell(row=6, column=cust_id_col, value="Customer ID")
    attr_names = list(config['attributes'].keys())
    for i, name in enumerate(attr_names):
        ws.cell(row=6, column=attr_start + i, value=name)
    ws.cell(row=6, column=cohort_col, value="Annual Cohort")

    # Period headers for section 1 (reference clean data)
    for i in range(num_dates):
        clean_col = clean_layout['arr_start'] + i
        ws.cell(row=6, column=s1_start + i,
                value=f"='{clean_sheet_name}'!{col_letter(clean_col)}6")

    # Period headers for section 2 (reference section 1, offset by 1)
    for i in range(num_dates - 1):
        ws.cell(row=6, column=s2_start + i,
                value=f"={col_letter(s1_start + i + 1)}6")

    # Period headers for section 3 (reference section 1)
    for i in range(num_dates):
        ws.cell(row=6, column=s3_start + i,
                value=f"={col_letter(s1_start + i)}6")

    # --- Data rows (Top N customers) ---
    first_customer_row = 7
    last_customer_row = first_customer_row + TOP_N - 1

    for rank in range(1, TOP_N + 1):
        row = first_customer_row + rank - 1

        # Rank number
        if rank == 1:
            ws.cell(row=row, column=rank_num_col, value=1)
        else:
            ws.cell(row=row, column=rank_num_col,
                    value=f"={col_letter(rank_num_col)}{row-1}+1")

        # Customer ID via XLOOKUP by rank
        rnl = col_letter(rank_num_col)
        ws.cell(row=row, column=cust_id_col,
                value=f"=_xlfn.XLOOKUP(${rnl}{row},"
                      f"'{clean_sheet_name}'!${clean_rank_col}${cdr_first}:${clean_rank_col}${cdr_last},"
                      f"'{clean_sheet_name}'!{col_letter(clean_layout['cust_id'])}${cdr_first}"
                      f":{col_letter(clean_layout['cust_id'])}${cdr_last})")

        # Attributes via XLOOKUP by rank
        for attr_idx in range(num_attrs):
            clean_attr_col = clean_layout['attr_start'] + attr_idx
            ws.cell(row=row, column=attr_start + attr_idx,
                    value=f"=_xlfn.XLOOKUP(${rnl}{row},"
                          f"'{clean_sheet_name}'!${clean_rank_col}${cdr_first}:${clean_rank_col}${cdr_last},"
                          f"'{clean_sheet_name}'!{col_letter(clean_attr_col)}${cdr_first}"
                          f":{col_letter(clean_attr_col)}${cdr_last})")

        # Cohort via XLOOKUP by rank
        ws.cell(row=row, column=cohort_col,
                value=f"=_xlfn.XLOOKUP(${rnl}{row},"
                      f"'{clean_sheet_name}'!${clean_rank_col}${cdr_first}:${clean_rank_col}${cdr_last},"
                      f"'{clean_sheet_name}'!{col_letter(clean_layout['cohort'])}${cdr_first}"
                      f":{col_letter(clean_layout['cohort'])}${cdr_last})")

        # Section 1: ARR per period (SUMIFS by customer ID)
        cidl = col_letter(cust_id_col)
        for i in range(num_dates):
            clean_col = clean_layout['arr_start'] + i
            ws.cell(row=row, column=s1_start + i,
                    value=f"=SUMIFS('{clean_sheet_name}'!{col_letter(clean_col)}${cdr_first}"
                          f":{col_letter(clean_col)}${cdr_last},"
                          f"'{clean_sheet_name}'!${clean_cust_id_col}${cdr_first}"
                          f":${clean_cust_id_col}${cdr_last},${cidl}{row})/{units_cell}")

        # Section 2: % YoY Growth
        for i in range(num_dates - 1):
            curr_col = col_letter(s1_start + i + 1)
            prev_col = col_letter(s1_start + i)
            ws.cell(row=row, column=s2_start + i,
                    value=f'=IFERROR({curr_col}{row}/{prev_col}{row}-1,"n.a.")')

        # Section 3: % of Total
        total_row = last_customer_row + 3  # skip Top N Total, Other
        for i in range(num_dates):
            arr_col = col_letter(s1_start + i)
            ws.cell(row=row, column=s3_start + i,
                    value=f"={arr_col}{row}/{arr_col}${total_row}")

    # --- Summary rows ---
    r_top_total = last_customer_row + 1
    r_other = r_top_total + 1
    r_total = r_other + 1

    # Top N Total
    ws.cell(row=r_top_total, column=cust_id_col, value=f"Top {TOP_N} Customers")
    for i in range(num_dates):
        arr_cl = col_letter(s1_start + i)
        ws.cell(row=r_top_total, column=s1_start + i,
                value=f"=SUM({arr_cl}{first_customer_row}:{arr_cl}{last_customer_row})")

        # Growth
        if i > 0:
            curr = col_letter(s1_start + i)
            prev = col_letter(s1_start + i - 1)
            ws.cell(row=r_top_total, column=s2_start + i - 1,
                    value=f'=IFERROR({curr}{r_top_total}/{prev}{r_top_total}-1,"n.a.")')

        # % of Total
        ws.cell(row=r_top_total, column=s3_start + i,
                value=f"={arr_cl}{r_top_total}/{arr_cl}${r_total}")

    # Other Customers = Total - Top N
    ws.cell(row=r_other, column=cust_id_col, value="(+) Other Customers")
    for i in range(num_dates):
        arr_cl = col_letter(s1_start + i)
        ws.cell(row=r_other, column=s1_start + i,
                value=f"={arr_cl}{r_total}-{arr_cl}{r_top_total}")

        if i > 0:
            curr = col_letter(s1_start + i)
            prev = col_letter(s1_start + i - 1)
            ws.cell(row=r_other, column=s2_start + i - 1,
                    value=f'=IFERROR({curr}{r_other}/{prev}{r_other}-1,"n.a.")')

        ws.cell(row=r_other, column=s3_start + i,
                value=f"={arr_cl}{r_other}/{arr_cl}${r_total}")

    # Total ARR (all customers)
    ws.cell(row=r_total, column=cust_id_col, value=f"Total {metric_label}")
    for i in range(num_dates):
        clean_col = clean_layout['arr_start'] + i
        ws.cell(row=r_total, column=s1_start + i,
                value=f'=SUMIFS(\'{clean_sheet_name}\'!{col_letter(clean_col)}${cdr_first}'
                      f':{col_letter(clean_col)}${cdr_last},'
                      f'\'{clean_sheet_name}\'!${clean_cust_id_col}${cdr_first}'
                      f':${clean_cust_id_col}${cdr_last},"<>")/{units_cell}')

        if i > 0:
            curr = col_letter(s1_start + i)
            prev = col_letter(s1_start + i - 1)
            ws.cell(row=r_total, column=s2_start + i - 1,
                    value=f'=IFERROR({curr}{r_total}/{prev}{r_total}-1,"n.a.")')

    # --- Concentration Memo ---
    r_memo_start = r_total + 2
    ws.cell(row=r_memo_start, column=cust_id_col, value="Memo:")

    for tier_idx, tier in enumerate(CONCENTRATION_TIERS):
        row = r_memo_start + 1 + tier_idx
        ws.cell(row=row, column=rank_num_col, value=tier)
        ws.cell(row=row, column=cust_id_col,
                value=f'="Top "&{col_letter(rank_num_col)}{row}&" Customers Concentration"')

        rnl = col_letter(rank_num_col)
        for i in range(num_dates):
            arr_cl = col_letter(s1_start + i)
            ws.cell(row=row, column=s1_start + i,
                    value=f"=SUMIFS({arr_cl}${first_customer_row}:{arr_cl}${last_customer_row},"
                          f"${rnl}${first_customer_row}:${rnl}${last_customer_row},"
                          f'"<="&${rnl}{row})')

            if i > 0:
                curr = col_letter(s1_start + i)
                prev = col_letter(s1_start + i - 1)
                ws.cell(row=row, column=s2_start + i - 1,
                        value=f'=IFERROR({curr}{row}/{prev}{row}-1,"n.a.")')

            ws.cell(row=row, column=s3_start + i,
                    value=f"={arr_cl}{row}/{arr_cl}${r_total}")

    return sheet_name

File: frontend/src/engine/top_customers.ts
/**
 * Top Customer Analysis tab generator.
 * Port of data-pack-app/engine/top_customers.py for client-side use with ExcelJS.
 */
import type { Workbook } from 'exceljs';
import type { EngineConfig } from './types';
import type { CleanLayout } from './utils';
import { colLetter } from './utils';

export const TOP_N = 25;
export const CONCENTRATION_TIERS = [3, 5, 10, 25];

export function generateTopCustomersTab(
  wb: Workbook, config: EngineConfig,
  cleanSheetName: string, cleanLayout: CleanLayout,
  firstDataRow: number, lastDataRow: number
): string {
  const sheetName = 'Annual Top Customer Analysis';
  const ws = wb.addWorksheet(sheetName);

  const numDates = cleanLayout.num_dates;
  const numAttrs = cleanLayout.num_attrs;
  const attrNames = Object.keys(config.attributes);

  const rankNumCol = 2;
  const custIdCol = 3;
  const attrStart = 4;
  const attrEnd = attrStart + numAttrs - 1;
  const cohortCol = attrEnd + 1;

  const s1Start = cohortCol + 1;
  const s1End = s1Start + numDates - 1;
  const s2Start = s1End + 2;
  const s2End = s2Start + numDates - 2;
  const s3Start = s2End + 2;
  const cdrFirst = firstDataRow;
  const cdrLast = lastDataRow;
  const cleanRankCol = colLetter(cleanLayout.rank);
  const cleanCustIdCol = colLetter(cleanLayout.cust_id);
  const metricLabel = (config.data_type || 'arr') === 'arr' ? 'ARR' : 'Revenue';

  // Row 3: Units
  ws.getCell(3, 1).value = 'Units';
  ws.getCell(3, rankNumCol).value = { formula: 'Control!$C$4' };
  const unitsCell = `$${colLetter(rankNumCol)}$3`;

  // Row 5: Section headers
  ws.getCell(5, s1Start).value = metricLabel;
  ws.getCell(5, s2Start).value = '% YoY Growth';
  ws.getCell(5, s3Start).value = '% of Total';

  // Row 6: Column headers
  ws.getCell(6, custIdCol).value = 'Customer ID';
  for (let i = 0; i < attrNames.length; i++) {
    ws.getCell(6, attrStart + i).value = attrNames[i];
  }
  ws.getCell(6, cohortCol).value = 'Annual Cohort';

  for (let i = 0; i < numDates; i++) {
    const cleanCol = cleanLayout.arr_start + i;
    ws.getCell(6, s1Start + i).value = { formula: `'${cleanSheetName}'!${colLetter(cleanCol)}6` };
  }
  for (let i = 0; i < numDates - 1; i++) {
    ws.getCell(6, s2Start + i).value = { formula: `${colLetter(s1Start + i + 1)}6` };
  }
  for (let i = 0; i < numDates; i++) {
    ws.getCell(6, s3Start + i).value = { formula: `${colLetter(s1Start + i)}6` };
  }

  // Data rows
  const firstCustomerRow = 7;
  const lastCustomerRow = firstCustomerRow + TOP_N - 1;

  for (let rank = 1; rank <= TOP_N; rank++) {
    const row = firstCustomerRow + rank - 1;

    if (rank === 1) {
      ws.getCell(row, rankNumCol).value = 1;
    } else {
      ws.getCell(row, rankNumCol).value = { formula: `${colLetter(rankNumCol)}${row - 1}+1` };
    }

    const rnl = colLetter(rankNumCol);
    ws.getCell(row, custIdCol).value = {
      formula: `_xlfn.XLOOKUP($${rnl}${row},'${cleanSheetName}'!$${cleanRankCol}$${cdrFirst}:$${cleanRankCol}$${cdrLast},'${cleanSheetName}'!${colLetter(cleanLayout.cust_id)}$${cdrFirst}:${colLetter(cleanLayout.cust_id)}$${cdrLast})`
    };

    for (let attrIdx = 0; attrIdx < numAttrs; attrIdx++) {
      const cleanAttrCol = cleanLayout.attr_start + attrIdx;
      ws.getCell(row, attrStart + attrIdx).value = {
        formula: `_xlfn.XLOOKUP($${rnl}${row},'${cleanSheetName}'!$${cleanRankCol}$${cdrFirst}:$${cleanRankCol}$${cdrLast},'${cleanSheetName}'!${colLetter(cleanAttrCol)}$${cdrFirst}:${colLetter(cleanAttrCol)}$${cdrLast})`
      };
    }

    ws.getCell(row, cohortCol).value = {
      formula: `_xlfn.XLOOKUP($${rnl}${row},'${cleanSheetName}'!$${cleanRankCol}$${cdrFirst}:$${cleanRankCol}$${cdrLast},'${cleanSheetName}'!${colLetter(cleanLayout.cohort)}$${cdrFirst}:${colLetter(cleanLayout.cohort)}$${cdrLast})`
    };

    // Section 1: ARR
    const cidl = colLetter(custIdCol);
    for (let i = 0; i < numDates; i++) {
      const cleanCol = cleanLayout.arr_start + i;
      ws.getCell(row, s1Start + i).value = {
        formula: `SUMIFS('${cleanSheetName}'!${colLetter(cleanCol)}$${cdrFirst}:${colLetter(cleanCol)}$${cdrLast},'${cleanSheetName}'!$${cleanCustIdCol}$${cdrFirst}:$${cleanCustIdCol}$${cdrLast},$${cidl}${row})/${unitsCell}`
      };
    }

    // Section 2: % YoY Growth
    for (let i = 0; i < numDates - 1; i++) {
      const currCol = colLetter(s1Start + i + 1);
      const prevCol = colLetter(s1Start + i);
      ws.getCell(row, s2Start + i).value = { formula: `IFERROR(${currCol}${row}/${prevCol}${row}-1,"n.a.")` };
    }

    // Section 3: % of Total
    const totalRow = lastCustomerRow + 3;
    for (let i = 0; i < numDates; i++) {
      const arrCol = colLetter(s1Start + i);
      ws.getCell(row, s3Start + i).value = { formula: `${arrCol}${row}/${arrCol}$${totalRow}` };
    }
  }

  // Summary rows
  const rTopTotal = lastCustomerRow + 1;
  const rOther = rTopTotal + 1;
  const rTotal = rOther + 1;

  ws.getCell(rTopTotal, custIdCol).value = `Top ${TOP_N} Customers`;
  for (let i = 0; i < numDates; i++) {
    const arrCl = colLetter(s1Start + i);
    ws.getCell(rTopTotal, s1Start + i).value = { formula: `SUM(${arrCl}${firstCustomerRow}:${arrCl}${lastCustomerRow})` };
    if (i > 0) {
      const curr = colLetter(s1Start + i);
      const prev = colLetter(s1Start + i - 1);
      ws.getCell(rTopTotal, s2Start + i - 1).value = { formula: `IFERROR(${curr}${rTopTotal}/${prev}${rTopTotal}-1,"n.a.")` };
    }
    ws.getCell(rTopTotal, s3Start + i).value = { formula: `${arrCl}${rTopTotal}/${arrCl}$${rTotal}` };
  }

  ws.getCell(rOther, custIdCol).value = '(+) Other Customers';
  for (let i = 0; i < numDates; i++) {
    const arrCl = colLetter(s1Start + i);
    ws.getCell(rOther, s1Start + i).value = { formula: `${arrCl}${rTotal}-${arrCl}${rTopTotal}` };
    if (i > 0) {
      const curr = colLetter(s1Start + i);
      const prev = colLetter(s1Start + i - 1);
      ws.getCell(rOther, s2Start + i - 1).value = { formula: `IFERROR(${curr}${rOther}/${prev}${rOther}-1,"n.a.")` };
    }
    ws.getCell(rOther, s3Start + i).value = { formula: `${arrCl}${rOther}/${arrCl}$${rTotal}` };
  }

  ws.getCell(rTotal, custIdCol).value = `Total ${metricLabel}`;
  for (let i = 0; i < numDates; i++) {
    const cleanCol = cleanLayout.arr_start + i;
    ws.getCell(rTotal, s1Start + i).value = {
      formula: `SUMIFS('${cleanSheetName}'!${colLetter(cleanCol)}$${cdrFirst}:${colLetter(cleanCol)}$${cdrLast},'${cleanSheetName}'!$${cleanCustIdCol}$${cdrFirst}:$${cleanCustIdCol}$${cdrLast},"<>")/${unitsCell}`
    };
    if (i > 0) {
      const curr = colLetter(s1Start + i);
      const prev = colLetter(s1Start + i - 1);
      ws.getCell(rTotal, s2Start + i - 1).value = { formula: `IFERROR(${curr}${rTotal}/${prev}${rTotal}-1,"n.a.")` };
    }
  }

  // Concentration Memo
  const rMemoStart = rTotal + 2;
  ws.getCell(rMemoStart, custIdCol).value = 'Memo:';

  for (let tierIdx = 0; tierIdx < CONCENTRATION_TIERS.length; tierIdx++) {
    const tier = CONCENTRATION_TIERS[tierIdx];
    const row = rMemoStart + 1 + tierIdx;
    ws.getCell(row, rankNumCol).value = tier;
    ws.getCell(row, custIdCol).value = { formula: `"Top "&${colLetter(rankNumCol)}${row}&" Customers Concentration"` };

    const rnl = colLetter(rankNumCol);
    for (let i = 0; i < numDates; i++) {
      const arrCl = colLetter(s1Start + i);
      ws.getCell(row, s1Start + i).value = {
        formula: `SUMIFS(${arrCl}$${firstCustomerRow}:${arrCl}$${lastCustomerRow},$${rnl}$${firstCustomerRow}:$${rnl}$${lastCustomerRow},"<="&$${rnl}${row})`
      };
      if (i > 0) {
        const curr = colLetter(s1Start + i);
        const prev = colLetter(s1Start + i - 1);
        ws.getCell(row, s2Start + i - 1).value = { formula: `IFERROR(${curr}${row}/${prev}${row}-1,"n.a.")` };
      }
      ws.getCell(row, s3Start + i).value = { formula: `${arrCl}${row}/${arrCl}$${rTotal}` };
    }
  }

  return sheetName;
}
